openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,358 @@
+    if (c != NULL) {
+        SSL_CTX_set_tlsext_status_arg(c->ctx, NULL);
+        SSL_CTX_set_tlsext_status_cb(c->ctx, NULL);
+    }
+
+    (*env)->DeleteWeakGlobalRef(env, callbackRef);
+}
+
+/**
+ * This function is being called from OpenSSL.
+ *
+ * See:
+ *     netty_ocsp_client_callback
+ *     netty_ocsp_server_callback
+ */
+static int netty_ocsp_callback(int error, SSL *ssl, void *arg) {
+
+    // NOTE: Some of the following error conditions are conceptually not
+    // possible and the checks exist just for completeness. The errors
+    // are being outputted to STDERR because the calling Thread is not
+    // attached to the JVM yet.
+
+    if (OCSP_CALLBACK_CLAZZ == NULL) {
+        fprintf(stderr, "netty_ocsp_callback: no callback class\n");
+        return error;
+    }
+
+    if (OCSP_CALLBACK_METHOD == NULL) {
+        fprintf(stderr, "netty_ocsp_callback: no callback method\n");
+        return error;
+    }
+
+    if (arg == NULL) {
+        fprintf(stderr, "netty_ocsp_callback: no callback object\n");
+        return error;
+    }
+
+    JavaVM *jvm = tcn_get_java_vm();
+    if (jvm == NULL) {
+        fprintf(stderr, "netty_ocsp_callback: no JVM\n");
+        return error;
+    }
+
+    JNIEnv *env;
+    if (tcn_get_java_env(&env) != JNI_OK) {
+        fprintf(stderr, "netty_ocsp_callback: no JNIEnv\n");
+        return error;
+    }
+
+    if ((*jvm)->AttachCurrentThread(jvm, (void **)&env, NULL) != JNI_OK) {
+        fprintf(stderr, "netty_ocsp_callback: AttachCurrentThread\n");
+        return error;
+    }
+
+    int response = 0;
+    jobject callback = (*env)->NewLocalRef(env, (jweak)arg);
+
+    if ((*env)->IsSameObject(env, callback, NULL) == JNI_FALSE) {
+
+        jlong ssl_ = P2J(ssl);
+        response = (int)(*env)->CallIntMethod(env, callback, OCSP_CALLBACK_METHOD, ssl_);
+
+        if ((*env)->ExceptionCheck(env)) {
+            // Use the given error code in the event of a JVM exception
+            response = error;
+
+            (*env)->ExceptionDescribe(env);
+            (*env)->ExceptionClear(env);
+        }
+    } else {
+        response = error;
+    }
+
+    (*env)->DeleteLocalRef(env, callback);
+    (*jvm)->DetachCurrentThread(jvm);
+
+    return response;
+}
+
+/**
+ * The client uses -1 for errors which is distinctly different from the server's error code.
+ *
+ * @see netty_ocsp_callback
+ */
+static int netty_ocsp_client_callback(SSL *ssl, void *arg) {
+    return netty_ocsp_callback(-1, ssl, arg);
+}
+
+/**
+ * The server uses SSL_TLSEXT_ERR_ALERT_FATAL for errors (it's a non-negative number).
+ *
+ * @see netty_ocsp_callback
+ */
+static int netty_ocsp_server_callback(SSL *ssl, void *arg) {
+    return netty_ocsp_callback(SSL_TLSEXT_ERR_ALERT_FATAL, ssl, arg);
+}
+
+
+/**
+ * Returns true is OCSP stapling is available.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSLContext, isOcspSupported)(TCN_STDARGS) {
+    return JNI_TRUE;
```

Comment:
The current `return true` is dependent on the outer `#ifdef` which says OpenSSL >= 1.0.2 and !defined(OPENSSL_NO_OCSP).